pkg/controllers/state/state.go

```diff
@@ -0,0 +1,241 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"knative.dev/pkg/logging"
+
+	"golang.org/x/time/rate"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	podutils "github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// Cluster maintains cluster state that is often needed but expensive to compute.
+type Cluster struct {
+	ctx         context.Context
+	rateLimiter *rate.Limiter
+	kubeClient  client.Client
+
+	// Pod Specific Tracking
+	knownPods        sync.Map // mapping of pod namespaced name to struct{} used to track if we've seen a pod before
```

Comment:
It might be a bit easier to reason able if we either use sync.Map for all of these or a mutex on the whole class, especially in cases where we're locking inside out a sync.Map `Range`. It's potentially higher performance to do what you're doing, but the RWLock should get us pretty far.

Reply:
With a sync map for the node states, we would need to deep copy the node, update the copy and replace it in the map to prevent races.

If we did this though, I would still want the mutex preserve the invariant that currently exists where bindings that we are aware of effect the resource utilization on the node.  With two maps, we could potentially have a binding that hasn't yet updated the node resource usage.